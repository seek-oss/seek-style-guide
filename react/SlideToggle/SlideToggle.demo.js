import SlideToggle from './SlideToggle';
import * as sketch from './SlideToggle.sketch';

export default {
  route: '/slidetoggle',
  title: 'Slide Toggle',
  category: 'Form',
  component: SlideToggle,
  sketch,
  initialProps: {
    onChange: () => {},
    id: 'myToggle',
    label: 'Toggle me'
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Hide label',
          transformProps: ({ className, ...props }) => ({
            ...props,
            hideLabel: true
          })
        }
      ]
    },
    {
      label: 'Label position',
      type: 'radio',
      states: [
        {
          label: 'Right',
          transformProps: props => ({
            ...props,
            position: 'right'
          })
        },
        {
          label: 'Left',
          transformProps: props => ({
            ...props,
            position: 'left'
          })
        }
      ]
    }
  ]
};
