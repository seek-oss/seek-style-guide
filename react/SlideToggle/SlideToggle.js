import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TickIcon from '../TickIcon/TickIcon';
import classnames from 'classnames';
import styles from './SlideToggle.less';

const LEFT = 'left';
const RIGHT = 'right';

export default class SlideToggle extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    position: PropTypes.oneOf([LEFT, RIGHT]),
    hideLabel: PropTypes.bool,
    children: PropTypes.node,
    checked: PropTypes.bool,
    className: PropTypes.string
  };

  static defaultProps = {
    position: RIGHT,
    hideLabel: false
  };

  renderLabel = currentPosition => {
    const { label, hideLabel, children, position } = this.props;
    const labelStyle = classnames(
      styles.labelText,
      position === LEFT ? styles.labelLeft : styles.labelRight
    );
    return (
      hideLabel !== true &&
      position === currentPosition &&
      !children && <span className={labelStyle}>{label}</span>
    );
  };

  render() {
    const { id, checked, label, position, onChange, className } = this.props;

    const inputStyles = classnames({
      [styles.input]: true,
      [styles.inputLeft]: position === LEFT
    });

    return (
      <div className={classnames(styles.root, className)}>
        <input
          id={id}
          ref={this.storeInputRef}
          type="checkbox"
          aria-label={label}
          className={inputStyles}
          checked={checked}
          onChange={onChange}
        />
        <label htmlFor={id} className={styles.switch}>
          {this.renderLabel(LEFT)}
          <div className={styles.slideContainer}>
            <div className={styles.slider} />
            <TickIcon
              className={styles.slideButton}
              svgClassName={styles.svg}
            />
          </div>
          {this.renderLabel(RIGHT)}
        </label>
      </div>
    );
  }
}
