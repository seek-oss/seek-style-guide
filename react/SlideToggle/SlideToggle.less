@import (reference) "~seek-style-guide/theme";

@toggle-height: (@row-height * 5);
@toggle-slider-width: 48px;
@toggle-slider-height: (@row-height * 4);
@toggle-button-border: 1px;
@toggle-radius: (@toggle-height / 2);
@slide-distance: @toggle-slider-width - (@toggle-height - @toggle-button-border);
@anticipation-distance: 4px;

.root {
  position: relative;
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  min-height: @touchableTextHeight;
  // IE11 Flex box centering fix
  height: 1px;
}

.switch {
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  user-select: none;

  // IE11 :active selector fix for nested elements to
  // allow the parent (the label) to accept the event
  // https://stackoverflow.com/questions/31061544/active-selector-in-ie11-doesnt-work-if-there-are-nested-elements-how-do-i-mak
  :global(*) {
    pointer-events: none;
  }
}

.input {
  opacity: 0;
  position: absolute;
  margin: 0;
  width: @toggle-slider-width;
  height: @toggle-height;
  cursor: pointer;
}

.labelText {
  .rawText(@standard-type-scale);
  line-height: @toggle-height;

  .input:checked ~ .switch & {
    font-weight: @sk-bold;
  }
}

.inputLeft {
  left: auto;
  right: 0;
}

.labelLeft {
  margin-right: @gutter-width / 2;
}

.labelRight {
  margin-left: @gutter-width / 2;
}

.slideContainer {
  width: @toggle-slider-width;
  height: @toggle-height;
  position: relative;
  flex-shrink: 0;
}

.slider {
  width: @toggle-slider-width;
  height: @toggle-slider-height;
  margin: ((@toggle-height - @toggle-slider-height) / 2) 0;
  border-radius: @toggle-radius;
  background-color: @sk-mid-gray-light;
  position: absolute;
  overflow: hidden;
  // Fix for Safari border-radius, overflow hidden, transform bug:
  // https://gist.github.com/ayamflow/b602ab436ac9f05660d9c15190f4fd7b
  -webkit-mask-image: -webkit-radial-gradient(white, black);

  &::before {
    content: "";
    position: absolute;
    height: 100%;
    width: 100%;
    background-color: @sk-form-accent;
    border-radius: @toggle-radius;
    transition: transform 0.3s ease;
    transform: translateX(-(@slide-distance + @anticipation-distance));
  }

  .input:checked ~ .switch &::before {
    transform: none;
  }
}

.slideButton {
  transition: transform 0.3s ease;
  height: @toggle-height;
  width: @toggle-height;
  border-radius: 50%;
  background-color: @sk-white;
  border: @toggle-button-border solid @sk-mid-gray-light;
  display: flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  transform: translateX(-@toggle-button-border);

  // Split the focus styles out to a pseudo element
  // to fix `box-shadow` rendering artifact on transition
  &::before {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    transition: opacity 0.3s ease;
    opacity: 0;
    .focus();
    border-radius: @toggle-radius;

    .input:not([readonly]):not([disabled]):focus ~ .switch &,
    .input:not([readonly]):not([disabled]):active ~ .switch & {
      opacity: 1;
    }
  }

  .input:active ~ .switch & {
    transform: translateX(-@anticipation-distance);
  }
  .input:checked ~ .switch & {
    transform: translateX(@slide-distance);
  }
  .input:checked:active ~ .switch & {
    transform: translateX(@slide-distance + @anticipation-distance);
  }
}

.svg {
  fill: @sk-form-accent;
  opacity: 0;
  transition: transform 0.3s ease, opacity 0.3s ease;
  transform: rotate(-25deg);

  .input:checked:active ~ .switch & {
    transform: rotate(6deg);
  }
  .input:checked ~ .switch & {
    transform: none;
    opacity: 1;
  }
}
