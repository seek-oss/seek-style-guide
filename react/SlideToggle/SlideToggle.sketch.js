import React from 'react';
import SlideToggle from './SlideToggle';

const noop = () => {};

export const symbols = {
  'Slide Toggle/Unchecked': (
    <SlideToggle
      id="myToggleUnchecked"
      label="Slide toggle"
      onChange={noop}
      checked={false}
    />
  ),
  'Slide Toggle/Checked': (
    <SlideToggle
      id="myToggleChecked"
      label="Slide toggle"
      onChange={noop}
      checked={true}
    />
  )
};
