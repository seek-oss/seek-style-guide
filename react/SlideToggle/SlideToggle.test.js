import React from 'react';
import { shallow } from 'enzyme';

import SlideToggle from './SlideToggle';

const renderSlideToggle = (props, children) =>
  shallow(
    <SlideToggle id="testSlideToggle" label="Test toggle" {...props}>
      {children}
    </SlideToggle>
  );

describe('Slide toggle:', () => {
  it('should render with default props', () => {
    expect(renderSlideToggle()).toMatchSnapshot();
  });

  it('should render a checked state', () => {
    expect(renderSlideToggle({ checked: true })).toMatchSnapshot();
  });

  it('should render with the label on the left', () => {
    expect(renderSlideToggle({ position: 'left' })).toMatchSnapshot();
  });

  it('should render with the label hidden', () => {
    expect(renderSlideToggle({ hideLabel: true })).toMatchSnapshot();
  });

  it('should call a passed in onChange function when the checkbox fires a change event', () => {
    const onChange = jest.fn();
    const slideToggle = renderSlideToggle({ onChange });
    const check = { target: { checked: true } };

    slideToggle.find('input').simulate('change', check);
    expect(onChange).toBeCalledWith(check);
  });
});
