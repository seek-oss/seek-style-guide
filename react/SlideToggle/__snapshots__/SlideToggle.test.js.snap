// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide toggle: should render a checked state 1`] = `
<div
  className="SlideToggle__root"
>
  <input
    aria-label="Test toggle"
    checked={true}
    className="SlideToggle__input"
    id="testSlideToggle"
    type="checkbox"
  />
  <label
    className="SlideToggle__switch"
    htmlFor="testSlideToggle"
  >
    <div
      className="SlideToggle__slideContainer"
    >
      <div
        className="SlideToggle__slider"
      />
      <TickIcon
        className="SlideToggle__slideButton"
        svgClassName="SlideToggle__svg"
      />
    </div>
    <span
      className="SlideToggle__labelText SlideToggle__labelRight"
    >
      Test toggle
    </span>
  </label>
</div>
`;

exports[`Slide toggle: should render with default props 1`] = `
<div
  className="SlideToggle__root"
>
  <input
    aria-label="Test toggle"
    className="SlideToggle__input"
    id="testSlideToggle"
    type="checkbox"
  />
  <label
    className="SlideToggle__switch"
    htmlFor="testSlideToggle"
  >
    <div
      className="SlideToggle__slideContainer"
    >
      <div
        className="SlideToggle__slider"
      />
      <TickIcon
        className="SlideToggle__slideButton"
        svgClassName="SlideToggle__svg"
      />
    </div>
    <span
      className="SlideToggle__labelText SlideToggle__labelRight"
    >
      Test toggle
    </span>
  </label>
</div>
`;

exports[`Slide toggle: should render with the label hidden 1`] = `
<div
  className="SlideToggle__root"
>
  <input
    aria-label="Test toggle"
    className="SlideToggle__input"
    id="testSlideToggle"
    type="checkbox"
  />
  <label
    className="SlideToggle__switch"
    htmlFor="testSlideToggle"
  >
    <div
      className="SlideToggle__slideContainer"
    >
      <div
        className="SlideToggle__slider"
      />
      <TickIcon
        className="SlideToggle__slideButton"
        svgClassName="SlideToggle__svg"
      />
    </div>
  </label>
</div>
`;

exports[`Slide toggle: should render with the label on the left 1`] = `
<div
  className="SlideToggle__root"
>
  <input
    aria-label="Test toggle"
    className="SlideToggle__input SlideToggle__inputLeft"
    id="testSlideToggle"
    type="checkbox"
  />
  <label
    className="SlideToggle__switch"
    htmlFor="testSlideToggle"
  >
    <span
      className="SlideToggle__labelText SlideToggle__labelLeft"
    >
      Test toggle
    </span>
    <div
      className="SlideToggle__slideContainer"
    >
      <div
        className="SlideToggle__slider"
      />
      <TickIcon
        className="SlideToggle__slideButton"
        svgClassName="SlideToggle__svg"
      />
    </div>
  </label>
</div>
`;
